### ISSUE TITLE
Failing to verify contract: "Cannot find module"

### ISSUE THREAD

**Issue by niccolopetti** [rkalis/truffle-plugin-verify#49]
```
D:\oracle\onchain>yarn run truffle run verify Oracle --network ropsten --debug
yarn run v1.22.4
$ D:\oracle\onchain\node_modules\.bin\truffle run verify Oracle --network ropsten --debug
DEBUG logging is turned ON
Running truffle-plugin-verify v0.5.2
Verifying Oracle
Reading artifact file at D:\oracle\onchain\build\contracts\Oracle.json
Retrieving constructor parameters from https://api-ropsten.etherscan.io/api?apiKey=VALIDAPIKEY&module=account&action=txlist&address=0x97879655Da10D93EaDA1a35665adeB94834A9fB3&page=1&sort=asc&offset=1
Constructor parameters retrieved: 0x
Cannot find module '/D/oracle/onchain/contracts/Oracle.sol'
Require stack:
- D:\oracle\onchain\node_modules\truffle-plugin-verify\verify.js
- D:\oracle\onchain\node_modules\original-require\index.js
- D:\oracle\onchain\node_modules\truffle\build\cli.bundled.js
Failed to verify 1 contract(s): Oracle

```
I'm trying to verify my contract but the plugin doesn't find it,

```
D:\oracle\onchain\contracts>ls
Migrations.sol  Oracle.sol
```
I am running on windows10 with node v12.18.0,
and the lastest versions of truffle and truffle-plugin-verify
How can I solve this?

**Comment by rkalis** [comment]
Hey, I had another report similar to this earlier, but then I didn't hear back from them so I couldn't look into it more. Could you share your artifact JSON file as well?

**Comment by niccolopetti** [comment]
Hey thanks for the answer, since the project I am working on is not open source can I send you the file in private or give you more specific details you may need? 
I am using solc 0.7.5 to compile

**Comment by niccolopetti** [comment]
I've sent you an email and I've tried to replace D:\oracle\onchain\build\contracts\Oracle.json with D:/oracle/onchain/contracts/Oracle.sol in the artifact and after that it found the module but I still got "Fail - Unable to verify"

**Comment by niccolopetti** [comment]
Yes that should work, drives are only single letters on windows and use the english alphabet

**Comment by rkalis** [comment]
Alright, I wrote some code that should fix this, but I don't want to deploy it just yet, cause I'm not sure it works and I don't use Windows. Could you install truffle-plugin-verify as a git-based dependency and see if it fixes it (and if not, please copy-paste the debug output again).

```
yarn add -D git://github.com/rkalis/truffle-plugin-verify.git#v0.5.3
```

**Comment by niccolopetti** [comment]
Ok I installed the new version and now it finds the module but I get
```
D:\oracle\onchain>yarn run truffle run verify Oracle --network ropsten --debug
yarn run v1.22.4
$ D:\oracle\onchain\node_modules\.bin\truffle run verify Oracle --network ropsten --debug
DEBUG logging is turned ON
Running truffle-plugin-verify v0.5.2
Verifying Oracle
Reading artifact file at D:\oracle\onchain\build\contracts\Oracle.json
Retrieving constructor parameters from https://api-ropsten.etherscan.io/api?apiKey=APIKEY&page=1&sort=asc&offset=1
Constructor parameters retrieved: 0x
Normalised contract path from /D/oracle/onchain/contracts/Oracle.sol to D:/oracle/onchain/contracts/Oracle.sol
Sending verify request with POST arguments:
{
  "apikey": "APIKEY",
  "module": "contract",
  "action": "verifysourcecode",
  "contractaddress": "ADDRESS",
  "sourceCode": "SOURCE CODE",
  "codeformat": "solidity-standard-json-input",
  "contractname": "D:/oracle/onchain/contracts/Oracle.sol:Oracle",
  "compilerversion": "v0.7.5+commit.eb77ed08",
  "constructorArguements": ""
}
Checking status of verification request cw2sbjifz3xkj1mzufr7xpankxlgrfc9ghis2gb7kns5ajapwq
Fail - Unable to verify
Failed to verify 1 contract(s): Oracle
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by rkalis** [comment]
Hmm alright. I'll look at this later. Maybe you could try checking out this repo and migrating + verifying the metacoin example. 

If that also fails then I know for sure that it is because of windows, but if that succeeds than it's specific to your project.

It's late over here, so I can post detailed instructions to run those examples later, but you might be able to figure that out by yourself as well.

**Comment by niccolopetti** [comment]
Ok I tried running the Metacoin example and it didn't work,
giving basically the same output as in my project, there might actually be a problem with windows.

**Comment by rkalis** [comment]
Alright, thanks for looking into it. I guess I'll have to run a Windows VM and do some debugging.

**Comment by niccolopetti** [comment]
You're welcome, thank you for working on it

**Comment by tokenDeveloper2020** [comment]
Hey, I am also facing same issue on Windows

**Comment by rkalis** [comment]
I did some debugging in a Windows VM today, and I got it working for the MetaCoin example. Could you install the new version from git (same branch) and try again?

**Comment by niccolopetti** [comment]
Ok I tried running the MetaCoin example as found in the project and it worked, but when I changed the solc version from 0.6.11 to 0.7.5 it didn't work

**Comment by tokenDeveloper2020** [comment]
Thank you, now its working with 0.7.0

**Comment by niccolopetti** [comment]
If it may be useful I tried migrating and verifying Migrations.sol and SimpleToken with solc 0.7.5 and it verified them, but Metacoin and my Smart Contract still fail to verify with 0.7.5

**Comment by rkalis** [comment]
I tried to verify the MetaCoin example from my MacBook using 0.7.5 and it failed as well, so it looks like this is a separate bug unrelated to Windows.

I then tried to use the Etherscan web form instead, which also failed. But it looks like the reason it failed is because somehow a part of the bytecode is replaced by some sort of placeholder.

The settings I used: https://goerli.etherscan.io/verifyContract-solc-json?a=0xf02e27c381EE14D68575a9Df553c67803709BCe8&c=v0.7.5%2bcommit.eb77ed08&lictype=1

With this standard JSON:

```json
{"language":"Solidity","sources":{"/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/ConvertLib.sol":{"content":"// SPDX-License-Identifier: MIT\npragma solidity >=0.4.25 <0.8.0;\n\nlibrary ConvertLib {\n\tstruct TestStruct {\n\t\tuint256 name;\n\t}\n\n\tfunction convert(uint amount,uint conversionRate) public pure returns (uint convertedAmount)\n\t{\n\t\treturn amount * conversionRate;\n\t}\n}\n"},"/Users/rosco/repos/ethereum/truffle-plugin-verify/test/metacoin/contracts/MetaCoin.sol":{"content":"// SPDX-License-Identifier: MIT\npragma experimental ABIEncoderV2;\npragma solidity >=0.4.25 <0.8.0;\n\nimport \"./ConvertLib.sol\";\n\n// This is just a simple example of a coin-like contract.\n// It is not standards compatible and cannot be expected to talk to other\n// coin/token contracts. If you want to create a standards-compliant\n// token, see: https://github.com/ConsenSys/Tokens. Cheers!\n\ncontract MetaCoin {\n\tmapping (address => uint) balances;\n\n\tevent Transfer(address indexed _from, address indexed _to, uint256 _value);\n\n\tconstructor() public {\n\t\tbalances[tx.origin] = 10000;\n\t}\n\n\tfunction sendCoin(address receiver, uint amount) public returns(bool sufficient) {\n\t\tif (balances[msg.sender] < amount) return false;\n\t\tbalances[msg.sender] -= amount;\n\t\tbalances[receiver] += amount;\n\t\temit Transfer(msg.sender, receiver, amount);\n\t\treturn true;\n\t}\n\n\tfunction getBalanceInEth(address addr) public view returns(uint){\n\t\treturn ConvertLib.convert(getBalance(addr),2);\n\t}\n\n\tfunction getBalance(address addr) public view returns(uint) {\n\t\treturn balances[addr];\n\t}\n\n\tfunction test(ConvertLib.TestStruct memory testObj) public pure {\n\t\trequire(testObj.name > 10);\n\t}\n}\n"}},"settings":{"remappings":[],"optimizer":{"enabled":true,"runs":200},"evmVersion":"byzantium","libraries":{"":{"ConvertLib":"0x6eeF5999c58142239aE618744434B5057151b2F6"}}}}
```

When I look at the output from Etherscan the entire bytecode is the same, except for a small part where the expected bytecode includes `6eef5999c58142239ae618744434b5057151b2f6` and the generated bytecode includes `__$5d5266654c55ddcb5675d52cef4063c43a$__` instead. My gut says this has something to do with library linking, but I can't be sure. In any case it appears that this should still be able to verify though, since the only difference in bytecode appears to be a placeholder.

What do you think @Enigmatic331, do you have an idea why this `__$...$__` is used? Likely something with SOlidity v0.7, but I couldn't find anything in the Solidity docs.

**Comment by rkalis** [comment]
I just did a few extra tests. Verification still works for 0.7.0, it doesn't for 0.8.0. @tokenDeveloper2020 can you check if it works for your project using 0.7.0. If so that could be a good workaround for now.

But it looks like some change made to Solidity between 0.7.0 and 0.7.5 makes it so that a part of the bytecode gets subbed out using `__$...$__` (I'm assuming it has something to do with library linking).

@chriseth if you have any information that could help out here that would be much appreciated (only this message and the previous one are relevant, the other messages can be ignored).

**Comment by Enigmatic331** [comment]
Hey @rkalis thanks for the head's up! I'm looking into this, will test through a couple versions. Will report back.

**Comment by rkalis** [comment]
I published v0.5.3 today which includes the bug fix for the "Cannot find module" bug. I'll leave this issue open for the other bug that we found.

**Comment by Enigmatic331** [comment]
Looks like v0.7.5 onward, library linking behaviour is different - Guessing it is something related to the "linking in assembly mode when library addresses are provided" enhancement.

I'm on this on our side. Using standard-json we expect the returned bytecode to have the library addresses to already be linked, so the change here I guess is to also perform library linking post-compilation. Working on it and will revert shortly.

**Comment by chriseth** [comment]
The `_$...$_` is indeed a placeholder for the library address. I cannot see any change being introduced in that regard, though. The changelog entry you referred to was only for assembly mode, not for solidity mode.

Speaking about linked binaries in general: It helps verification on other platforms if linking is performed by the compiler instead of manually substituting the placeholders by the library address. In that case, no placeholders should be present, except in the case that the library name could not be found - maybe this is the problem here?

**Comment by Enigmatic331** [comment]
Aha... Chris! <3

So looks like the libraries json object is what is causing it:

```
      "libraries":{
         "":{
            "ConvertLib":"0x6eeF5999c58142239aE618744434B5057151b2F6"
         }
      }
```

If it's the below, no issues for solc v0.7.5:
```
      "libraries":{
         "ConvertLib.sol":{
            "ConvertLib":"0x6eeF5999c58142239aE618744434B5057151b2F6"
         }
      }

```

solc v0.7.4 standardjson input supports both these versions, whereas solc v0.7.5 prefers only the latter example.

Hmmm. @rkalis where did you get your json generated from? I'll also check on my end as I am aware Etherscan does convert files to standard-json prior to compilation - So to make sure we are alright here too.

**Comment by rkalis** [comment]
In general it's from the metadata in Truffle's artifact JSON file, but the library object I actually hack in there, because it wasn't available from the metadata field.

**Comment by Enigmatic331** [comment]
Let's see... Yeap, looks right - That's also what I tried and seems to work (also test verified on your contract above: https://goerli.etherscan.io/address/0xf02e27c381EE14D68575a9Df553c67803709BCe8#code).

**Comment by rkalis** [comment]
Great! Thanks for the help everyone in this thread! I just published v0.5.4 that fixes this library linking bug for Solidity >=0.7.5. It's a bit convoluted since the truffle artifact doesn't store the linked libraries' source paths, so I had to read the libraries' artifact files and get the required information from there. I've verified that it works for the MetaCoin + ConvertLib example on Max and Windows.

I'm marking this issue as closed, but @niccolopetti please verify if verification is working for your project and feel free to re-open this issue or open a new issue if it still isn't working.

**Comment by niccolopetti** [comment]
It works, great job!
